File: porycon2/porycon/animation_parser.py
"""
Parser for Pokemon Emerald animation metadata from source code.
"""

import re
from pathlib import Path
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, field
from .logging_config import get_logger

logger = get_logger('animation_parser')


@dataclass
class AnimFrame:
    """Single frame in an animation sequence."""
    frame_index: int
    duration: int
    flip_horizontal: bool = False


@dataclass
class AnimationDefinition:
    """Animation definition with name and frames."""
    name: str
    frames: List[AnimFrame] = field(default_factory=list)


@dataclass
class SpriteAnimationMetadata:
    """Metadata for a sprite's animations."""
    sprite_name: str
    animation_table: str
    logical_frame_count: int
    physical_frame_mapping: Optional[List[int]] = None
    animation_definitions: List[AnimationDefinition] = field(default_factory=list)


@dataclass
class SpriteSourceInfo:
    """Information about a sprite source file."""
    pic_name: str
    file_path: str
    start_frame: int
    frame_count: int


class PokeemeraldAnimationParser:
    """Parses animation metadata from pokeemerald source code."""
    
    def __init__(self, pokeemerald_path: str):
        """
        Initialize parser.
        
        Args:
            pokeemerald_path: Path to pokeemerald root directory
        """
        self.pokeemerald_path = Path(pokeemerald_path)
        self.filename_mapping: Dict[str, str] = {}
    
    def get_filename_mapping(self) -> Dict[str, str]:
        """Get the filename to sprite name mapping."""
        return self.filename_mapping
    
    def parse_animation_data(self) -> Dict[str, SpriteAnimationMetadata]:
        """
        Parse animation data from pokeemerald source.
        
        Returns:
            Dictionary mapping sprite name to animation metadata
        """
        logger.info("Parsing animation data from pokeemerald source...")
        
        metadata: Dict[str, SpriteAnimationMetadata] = {}
        
        # Paths to source files
        graphics_info_path = self.pokeemerald_path / "src" / "data" / "object_events" / "object_event_graphics_info.h"
        graphics_path = self.pokeemerald_path / "src" / "data" / "object_events" / "object_event_graphics.h"
        pic_tables_path = self.pokeemerald_path / "src" / "data" / "object_events" / "object_event_pic_tables.h"
        anims_path = self.pokeemerald_path / "src" / "data" / "object_events" / "object_event_anims.h"
        
        if not all(p.exists() for p in [graphics_info_path, pic_tables_path, anims_path]):
            logger.warning("Could not find pokeemerald source files. Using default animations.")
            return metadata
        
        # Parse filename -> sprite name mapping
        filename_mapping = self._parse_filename_mapping(graphics_path)
        
        # Parse animation tables
        animation_tables = self._parse_animation_tables(anims_path)
        
        # Parse sprite graphics info
        graphics_info = self._parse_graphics_info(graphics_info_path)
        
        # Parse frame counts and mappings from pic tables
        frame_counts = self._parse_frame_counts(pic_tables_path)
        frame_mappings = self._parse_frame_mappings(pic_tables_path)
        
        # Combine the data
        for sprite_name, anim_table in graphics_info.items():
            if sprite_name in frame_counts:
                frame_count = frame_counts[sprite_name]
                metadata[sprite_name] = SpriteAnimationMetadata(
                    sprite_name=sprite_name,
                    animation_table=anim_table,
                    logical_frame_count=frame_count,
                    physical_frame_mapping=frame_mappings.get(sprite_name),
                    animation_definitions=animation_tables.get(anim_table, [])
                )
        
        logger.info(f"Parsed animation data for {len(metadata)} sprites")
        logger.info(f"Parsed {len(filename_mapping)} filename->sprite mappings")
        
        # Store filename mapping
        self.filename_mapping = filename_mapping
        
        return metadata
    
    def _parse_filename_mapping(self, file_path: Path) -> Dict[str, str]:
        """
        Parse filename -> sprite name mapping from object_event_graphics.h
        
        Returns:
            Dictionary mapping "directory/filename" -> sprite name
        """
        result: Dict[str, str] = {}
        if not file_path.exists():
            return result
        
        content = file_path.read_text(encoding='utf-8')
        
        # Match: const u32 gObjectEventPic_<SpriteName>[] = INCBIN_U32("graphics/object_events/pics/people/<path>/<filename>.4bpp");
        pattern = re.compile(
            r'gObjectEventPic_(\w+)\[\]\s*=\s*INCBIN_U32\("graphics/object_events/pics/people/([^"]+)\.4bpp"\)'
        )
        
        for match in pattern.finditer(content):
            sprite_name = match.group(1)  # e.g. "BrendanNormal", "MayRunning"
            sprite_file_path = match.group(2)  # e.g. "brendan/walking", "may/running"
            
            # Extract just the filename without path
            path_obj = Path(sprite_file_path)
            filename = path_obj.name  # e.g. "walking", "running"
            directory = str(path_obj.parent).replace("\\", "/")  # e.g. "brendan", "may"
            
            # Create key as "directory/filename" to match our extraction
            key = f"{directory}/{filename}" if directory else filename
            result[key] = sprite_name
            
            logger.debug(f"  Filename mapping: {key} -> {sprite_name}")
        
        return result
    
    def parse_pic_table_sources(
        self,
        pic_tables_path: Path,
        graphics_path: Path
    ) -> Dict[str, List[SpriteSourceInfo]]:
        """
        Parse which PNG files belong to which sPicTable.
        
        Returns:
            Dictionary mapping sPicTable name -> list of SpriteSourceInfo
        """
        result: Dict[str, List[SpriteSourceInfo]] = {}
        
        if not pic_tables_path.exists() or not graphics_path.exists():
            return result
        
        pic_tables_content = pic_tables_path.read_text(encoding='utf-8')
        graphics_content = graphics_path.read_text(encoding='utf-8')
        
        # First, build a map of gObjectEventPic names to file paths
        pic_to_file: Dict[str, str] = {}
        file_pattern = re.compile(
            r'gObjectEventPic_(\w+)\[\]\s*=\s*INCBIN_U32\("graphics/object_events/pics/people/([^"]+)\.4bpp"\)'
        )
        
        for match in file_pattern.finditer(graphics_content):
            pic_to_file[match.group(1)] = match.group(2)
        
        # Parse each sPicTable and find which gObjectEventPic_* it references
        table_pattern = re.compile(
            r'sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}',
            re.DOTALL
        )
        
        for table_match in table_pattern.finditer(pic_tables_content):
            table_name = table_match.group(1)
            table_content = table_match.group(2)
            
            sources: List[SpriteSourceInfo] = []
            seen_pics: Dict[str, SpriteSourceInfo] = {}
            
            # Find all gObjectEventPic references in this table
            frame_pattern = re.compile(
                r'(?:overworld_frame|obj_frame_tiles)\(gObjectEventPic_(\w+)'
            )
            
            for frame_match in frame_pattern.finditer(table_content):
                pic_name = frame_match.group(1)
                
                if pic_name not in seen_pics:
                    source_info = SpriteSourceInfo(
                        pic_name=pic_name,
                        file_path=pic_to_file.get(pic_name, ""),
                        start_frame=sum(s.frame_count for s in seen_pics.values()),
                        frame_count=0
                    )
                    seen_pics[pic_name] = source_info
                    sources.append(source_info)
                
                seen_pics[pic_name].frame_count += 1
            
            if sources:
                result[table_name] = sources
        
        return result
    
    def _parse_graphics_info(self, file_path: Path) -> Dict[str, str]:
        """
        Parse sprite graphics info to get animation table assignments.
        
        Returns:
            Dictionary mapping sprite name -> animation table name
        """
        result: Dict[str, str] = {}
        if not file_path.exists():
            return result
        
        content = file_path.read_text(encoding='utf-8')
        
        # Match: const struct ObjectEventGraphicsInfo gObjectEventGraphicsInfo_<Name> = { ... .anims = sAnimTable_<AnimTable>, ... }
        pattern = re.compile(
            r'gObjectEventGraphicsInfo_(\w+)\s*=\s*\{[^}]*\.anims\s*=\s*sAnimTable_(\w+)',
            re.DOTALL
        )
        
        for match in pattern.finditer(content):
            sprite_name = match.group(1)
            anim_table = match.group(2)
            result[sprite_name] = anim_table
        
        return result
    
    def _parse_frame_counts(self, file_path: Path) -> Dict[str, int]:
        """
        Parse frame counts from pic tables.
        
        Returns:
            Dictionary mapping sprite name -> frame count
        """
        result: Dict[str, int] = {}
        if not file_path.exists():
            return result
        
        content = file_path.read_text(encoding='utf-8')
        
        # Match: sPicTable_<Name>[] = { ... }
        pattern = re.compile(r'sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}', re.DOTALL)
        
        for match in pattern.finditer(content):
            sprite_name = match.group(1)
            table_content = match.group(2)
            
            # Count the number of overworld_frame or obj_frame_tiles entries
            frame_matches = re.findall(r'overworld_frame|obj_frame_tiles', table_content)
            result[sprite_name] = len(frame_matches)
        
        return result
    
    def _parse_frame_mappings(self, file_path: Path) -> Dict[str, List[int]]:
        """
        Parse physical frame mappings from pic tables.
        
        CRITICAL FIX: When a pic table references multiple PNG files (e.g., walking.png + running.png),
        we need to adjust frame indices based on which PNG they come from:
        - Frames from walking.png (first PNG): use frame_index as-is (0-8)
        - Frames from running.png (second PNG): add offset of 9 (walking frame count) to get (9-17)
        
        Returns:
            Dictionary mapping sprite name -> list of physical frame indices in the combined spritesheet
        """
        result: Dict[str, List[int]] = {}
        if not file_path.exists():
            return result
        
        content = file_path.read_text(encoding='utf-8')
        
        # First, parse pic table sources to know frame offsets
        graphics_path = file_path.parent / "object_event_graphics.h"
        pic_table_sources = self.parse_pic_table_sources(file_path, graphics_path)
        
        # Match: sPicTable_<Name>[] = { ... }
        pattern = re.compile(r'sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}', re.DOTALL)
        
        for match in pattern.finditer(content):
            sprite_name = match.group(1)
            table_content = match.group(2)
            
            # Get source info for this sprite (to know frame offsets)
            sources = pic_table_sources.get(sprite_name, [])
            
            # Build a map: pic_name -> start_frame_offset
            pic_to_offset: Dict[str, int] = {}
            for source in sources:
                pic_to_offset[source.pic_name] = source.start_frame
            
            # Parse overworld_frame(gObjectEventPic_<PicName>, ..., frame_index_in_png)
            frame_mapping: List[int] = []
            frame_pattern = re.compile(r'overworld_frame\(gObjectEventPic_(\w+),\s*\d+,\s*\d+,\s*(\d+)\)')
            
            for frame_match in frame_pattern.finditer(table_content):
                pic_name = frame_match.group(1)
                frame_index_in_png = int(frame_match.group(2))
                
                # Adjust frame index based on which PNG this came from
                offset = pic_to_offset.get(pic_name, 0)
                physical_frame_index = offset + frame_index_in_png
                frame_mapping.append(physical_frame_index)
            
            # Handle obj_frame_tiles (single frame sprites)
            if not frame_mapping and 'obj_frame_tiles' in table_content:
                frame_mapping.append(0)  # Single frame, index 0
            
            if frame_mapping:
                result[sprite_name] = frame_mapping
        
        return result
    
    def _parse_animation_tables(self, file_path: Path) -> Dict[str, List[AnimationDefinition]]:
        """
        Parse animation tables from object_event_anims.h.
        
        Returns:
            Dictionary mapping animation table name -> list of AnimationDefinition
        """
        result: Dict[str, List[AnimationDefinition]] = {}
        if not file_path.exists():
            return result
        
        content = file_path.read_text(encoding='utf-8')
        
        # Parse individual animation sequences
        anim_sequences: Dict[str, List[AnimFrame]] = {}
        seq_pattern = re.compile(r'sAnim_(\w+)\[\]\s*=\s*\{([^}]+)\}', re.DOTALL)
        
        for match in seq_pattern.finditer(content):
            anim_name = match.group(1)
            anim_content = match.group(2)
            frames: List[AnimFrame] = []
            
            # Parse ANIMCMD_FRAME(frameIndex, duration, .hFlip = true/false)
            frame_pattern = re.compile(
                r'ANIMCMD_FRAME\((\d+),\s*(\d+)(?:,\s*\.hFlip\s*=\s*(TRUE|FALSE))?\)'
            )
            
            for frame_match in frame_pattern.finditer(anim_content):
                frames.append(AnimFrame(
                    frame_index=int(frame_match.group(1)),
                    duration=int(frame_match.group(2)),
                    flip_horizontal=frame_match.group(3) == "TRUE"
                ))
            
            if frames:
                anim_sequences[f"sAnim_{anim_name}"] = frames
        
        # Parse animation tables that reference sequences
        table_pattern = re.compile(
            r'sAnimTable_(\w+)\[\]\s*=\s*\{([^}]+)\}',
            re.DOTALL
        )
        
        for match in table_pattern.finditer(content):
            table_name = match.group(1)
            table_content = match.group(2)
            animations: List[AnimationDefinition] = []
            
            # Parse [ANIM_*] = sAnim_<Name>,
            # Matches both ANIM_STD_* and other patterns
            entry_pattern = re.compile(r'\[ANIM_(?:STD_)?([A-Z_]+)\]\s*=\s*sAnim_(\w+)')
            
            for entry_match in entry_pattern.finditer(table_content):
                anim_type = entry_match.group(1)
                anim_seq_name = f"sAnim_{entry_match.group(2)}"
                
                if anim_seq_name in anim_sequences:
                    animations.append(AnimationDefinition(
                        name=self._convert_anim_type_name(anim_type),
                        frames=anim_sequences[anim_seq_name]
                    ))
            
            if animations:
                result[table_name] = animations
        
        return result
    
    def _convert_anim_type_name(self, emerald_name: str) -> str:
        """
        Convert pokeemerald's SCREAMING_SNAKE_CASE to lowercase_snake_case.
        
        Examples:
            FACE_SOUTH -> face_south
            GO_SOUTH -> go_south
            TAKE_OUT_ROD_SOUTH -> take_out_rod_south
        """
        return emerald_name.lower()

File: Porycon3/Services/AnimationParser.cs
using System.Text.RegularExpressions;

namespace Porycon3.Services;

/// <summary>
/// Parses animation metadata from pokeemerald source code.
/// </summary>
public class AnimationParser
{
    private readonly string _pokeemeraldPath;
    private readonly bool _verbose;

    public AnimationParser(string pokeemeraldPath, bool verbose = false)
    {
        _pokeemeraldPath = pokeemeraldPath;
        _verbose = verbose;
    }

    /// <summary>
    /// Parse all animation data from pokeemerald source.
    /// </summary>
    public AnimationData ParseAnimationData()
    {
        var result = new AnimationData();

        var objectEventsPath = Path.Combine(_pokeemeraldPath, "src", "data", "object_events");
        var graphicsInfoPath = Path.Combine(objectEventsPath, "object_event_graphics_info.h");
        var graphicsPath = Path.Combine(objectEventsPath, "object_event_graphics.h");
        var picTablesPath = Path.Combine(objectEventsPath, "object_event_pic_tables.h");
        var animsPath = Path.Combine(objectEventsPath, "object_event_anims.h");

        if (!Directory.Exists(objectEventsPath))
        {
            if (_verbose)
                Console.WriteLine($"[AnimationParser] Object events path not found: {objectEventsPath}");
            return result;
        }

        // Parse pic name -> file path mapping from object_event_graphics.h
        result.PicToFilePath = ParsePicToFilePath(graphicsPath);
        result.MultiFilePics = new Dictionary<string, List<string>>(MultiFilePics);

        // Parse animation sequences from object_event_anims.h
        result.AnimationSequences = ParseAnimationSequences(animsPath);
        result.AnimationTables = ParseAnimationTables(animsPath);

        // Parse sprite -> animation table mapping from object_event_graphics_info.h
        result.SpriteToAnimTable = ParseGraphicsInfo(graphicsInfoPath);

        // Parse pic table sources (which PNGs belong to which sprite)
        result.PicTableSources = ParsePicTableSources(picTablesPath);

        // Parse frame counts from pic tables
        result.FrameCounts = ParseFrameCounts(picTablesPath);

        // Parse physical frame mappings
        result.FrameMappings = ParseFrameMappings(picTablesPath, result.PicTableSources);

        if (_verbose)
        {
            Console.WriteLine($"[AnimationParser] Parsed {result.PicToFilePath.Count} pic->file mappings");
            Console.WriteLine($"[AnimationParser] Parsed {result.AnimationSequences.Count} animation sequences");
            Console.WriteLine($"[AnimationParser] Parsed {result.AnimationTables.Count} animation tables");
            Console.WriteLine($"[AnimationParser] Parsed {result.PicTableSources.Count} pic table sources");
        }

        return result;
    }

    /// <summary>
    /// Parse gObjectEventPic_* -> file path mapping.
    /// Handles both single and multi-file INCBIN_U32.
    /// </summary>
    private Dictionary<string, string> ParsePicToFilePath(string filePath)
    {
        var result = new Dictionary<string, string>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);

        // Match: gObjectEventPic_<Name>[] = INCBIN_U32("...", "...", ...);
        // Captures the pic name and the entire INCBIN content
        var pattern = new Regex(
            @"gObjectEventPic_(\w+)\[\]\s*=\s*INCBIN_U32\(([^)]+)\)",
            RegexOptions.Compiled);

        // Pattern to extract individual file paths from INCBIN content
        // Matches all subdirectories: people, berry_trees, cushions, dolls, misc, pokemon_old
        var filePattern = new Regex(
            @"""graphics/object_events/pics/([^""]+)\.4bpp""",
            RegexOptions.Compiled);

        foreach (Match match in pattern.Matches(content))
        {
            var picName = match.Groups[1].Value;
            var incbinContent = match.Groups[2].Value;

            // Extract all file paths from this INCBIN
            var fileMatches = filePattern.Matches(incbinContent);
            if (fileMatches.Count > 0)
            {
                // Store first file as primary (for single-file compat)
                result[picName] = fileMatches[0].Groups[1].Value;

                // Store multi-file mapping separately
                if (fileMatches.Count > 1)
                {
                    var allFiles = fileMatches.Select(m => m.Groups[1].Value).ToList();
                    MultiFilePics[picName] = allFiles;
                }
            }
        }

        return result;
    }

    /// <summary>
    /// Pics that combine multiple source files.
    /// </summary>
    public Dictionary<string, List<string>> MultiFilePics { get; } = new();

    /// <summary>
    /// Parse individual animation sequences (sAnim_*).
    /// </summary>
    private Dictionary<string, List<AnimFrame>> ParseAnimationSequences(string filePath)
    {
        var result = new Dictionary<string, List<AnimFrame>>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);

        // Match: sAnim_<Name>[] = { ... }
        var seqPattern = new Regex(
            @"sAnim_(\w+)\[\]\s*=\s*\{([^}]+)\}",
            RegexOptions.Compiled | RegexOptions.Singleline);

        // Match: ANIMCMD_FRAME(frameIndex, duration, .hFlip = TRUE/FALSE)
        var framePattern = new Regex(
            @"ANIMCMD_FRAME\((\d+),\s*(\d+)(?:,\s*\.hFlip\s*=\s*(TRUE|FALSE))?\)",
            RegexOptions.Compiled);

        foreach (Match seqMatch in seqPattern.Matches(content))
        {
            var animName = $"sAnim_{seqMatch.Groups[1].Value}";
            var animContent = seqMatch.Groups[2].Value;
            var frames = new List<AnimFrame>();

            foreach (Match frameMatch in framePattern.Matches(animContent))
            {
                frames.Add(new AnimFrame
                {
                    FrameIndex = int.Parse(frameMatch.Groups[1].Value),
                    Duration = int.Parse(frameMatch.Groups[2].Value),
                    FlipHorizontal = frameMatch.Groups[3].Value == "TRUE"
                });
            }

            if (frames.Count > 0)
            {
                result[animName] = frames;
            }
        }

        return result;
    }

    /// <summary>
    /// Parse animation tables (sAnimTable_*) that reference sequences.
    /// </summary>
    private Dictionary<string, List<SpriteAnimationDefinition>> ParseAnimationTables(string filePath)
    {
        var result = new Dictionary<string, List<SpriteAnimationDefinition>>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);
        var sequences = ParseAnimationSequences(filePath);

        // Match: sAnimTable_<Name>[] = { ... }
        var tablePattern = new Regex(
            @"sAnimTable_(\w+)\[\]\s*=\s*\{([^}]+)\}",
            RegexOptions.Compiled | RegexOptions.Singleline);

        // Match: [ANIM_*] = sAnim_<Name>,
        var entryPattern = new Regex(
            @"\[ANIM_(?:STD_)?([A-Z_]+)\]\s*=\s*sAnim_(\w+)",
            RegexOptions.Compiled);

        foreach (Match tableMatch in tablePattern.Matches(content))
        {
            var tableName = tableMatch.Groups[1].Value;
            var tableContent = tableMatch.Groups[2].Value;
            var animations = new List<SpriteAnimationDefinition>();

            foreach (Match entryMatch in entryPattern.Matches(tableContent))
            {
                var animType = entryMatch.Groups[1].Value.ToLowerInvariant();
                var seqName = $"sAnim_{entryMatch.Groups[2].Value}";

                if (sequences.TryGetValue(seqName, out var frames))
                {
                    animations.Add(new SpriteAnimationDefinition
                    {
                        Name = animType,
                        Frames = frames
                    });
                }
            }

            if (animations.Count > 0)
            {
                result[tableName] = animations;
            }
        }

        return result;
    }

    /// <summary>
    /// Parse sprite -> animation table mapping from graphics_info.h
    /// </summary>
    private Dictionary<string, string> ParseGraphicsInfo(string filePath)
    {
        var result = new Dictionary<string, string>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);

        // Match: gObjectEventGraphicsInfo_<Name> = { ... .anims = sAnimTable_<AnimTable>, ... }
        var pattern = new Regex(
            @"gObjectEventGraphicsInfo_(\w+)\s*=\s*\{[^}]*\.anims\s*=\s*sAnimTable_(\w+)",
            RegexOptions.Compiled | RegexOptions.Singleline);

        foreach (Match match in pattern.Matches(content))
        {
            result[match.Groups[1].Value] = match.Groups[2].Value;
        }

        return result;
    }

    /// <summary>
    /// Parse which PNGs belong to which sPicTable.
    /// </summary>
    private Dictionary<string, List<SpriteSourceInfo>> ParsePicTableSources(string filePath)
    {
        var result = new Dictionary<string, List<SpriteSourceInfo>>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);
        var graphicsPath = Path.Combine(Path.GetDirectoryName(filePath)!, "object_event_graphics.h");
        var picToFile = ParsePicToFilePath(graphicsPath);

        // Match: sPicTable_<Name>[] = { ... }
        var tablePattern = new Regex(
            @"sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}",
            RegexOptions.Compiled | RegexOptions.Singleline);

        // Match: overworld_frame, overworld_ascending_frames, or obj_frame_tiles references
        var framePattern = new Regex(
            @"(?:overworld_frame|overworld_ascending_frames|obj_frame_tiles)\(gObjectEventPic_(\w+)",
            RegexOptions.Compiled);

        foreach (Match tableMatch in tablePattern.Matches(content))
        {
            var tableName = tableMatch.Groups[1].Value;
            var tableContent = tableMatch.Groups[2].Value;
            var sources = new List<SpriteSourceInfo>();
            var seenPics = new Dictionary<string, SpriteSourceInfo>();

            foreach (Match frameMatch in framePattern.Matches(tableContent))
            {
                var picName = frameMatch.Groups[1].Value;

                if (!seenPics.ContainsKey(picName))
                {
                    var sourceInfo = new SpriteSourceInfo
                    {
                        PicName = picName,
                        FilePath = picToFile.GetValueOrDefault(picName, ""),
                        StartFrame = seenPics.Values.Sum(s => s.FrameCount),
                        FrameCount = 0
                    };
                    seenPics[picName] = sourceInfo;
                    sources.Add(sourceInfo);
                }

                seenPics[picName].FrameCount++;
            }

            if (sources.Count > 0)
            {
                result[tableName] = sources;
            }
        }

        return result;
    }

    /// <summary>
    /// Parse frame counts from pic tables.
    /// </summary>
    private Dictionary<string, int> ParseFrameCounts(string filePath)
    {
        var result = new Dictionary<string, int>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);

        // Match: sPicTable_<Name>[] = { ... }
        var pattern = new Regex(
            @"sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}",
            RegexOptions.Compiled | RegexOptions.Singleline);

        foreach (Match match in pattern.Matches(content))
        {
            var spriteName = match.Groups[1].Value;
            var tableContent = match.Groups[2].Value;

            // Count individual frames
            var frameMatches = Regex.Matches(tableContent, @"overworld_frame|obj_frame_tiles");
            var frameCount = frameMatches.Count;

            // Handle overworld_ascending_frames - these generate 9 or 18 frames
            // based on whether it's a normal (9) or combined walk+run sprite (18)
            var ascendingMatches = Regex.Matches(tableContent, @"overworld_ascending_frames\(gObjectEventPic_(\w+)");
            foreach (Match ascMatch in ascendingMatches)
            {
                var picName = ascMatch.Groups[1].Value;
                // Check if this is a multi-file pic (walking + running = 18 frames)
                if (MultiFilePics.ContainsKey(picName))
                {
                    frameCount += 18; // 9 walking + 9 running frames
                }
                else
                {
                    frameCount += 9; // Standard 9 frames
                }
            }

            result[spriteName] = frameCount;
        }

        return result;
    }

    /// <summary>
    /// Parse physical frame mappings from pic tables.
    /// </summary>
    private Dictionary<string, List<int>> ParseFrameMappings(
        string filePath,
        Dictionary<string, List<SpriteSourceInfo>> picTableSources)
    {
        var result = new Dictionary<string, List<int>>();
        if (!File.Exists(filePath)) return result;

        var content = File.ReadAllText(filePath);

        // Match: sPicTable_<Name>[] = { ... }
        var tablePattern = new Regex(
            @"sPicTable_(\w+)\[\]\s*=\s*\{([^}]+)\}",
            RegexOptions.Compiled | RegexOptions.Singleline);

        // Match: overworld_frame(gObjectEventPic_<PicName>, width, height, frame_index)
        var framePattern = new Regex(
            @"overworld_frame\(gObjectEventPic_(\w+),\s*\d+,\s*\d+,\s*(\d+)\)",
            RegexOptions.Compiled);

        foreach (Match tableMatch in tablePattern.Matches(content))
        {
            var spriteName = tableMatch.Groups[1].Value;
            var tableContent = tableMatch.Groups[2].Value;

            // Build pic name -> offset mapping
            var picToOffset = new Dictionary<string, int>();
            if (picTableSources.TryGetValue(spriteName, out var sources))
            {
                foreach (var source in sources)
                {
                    picToOffset[source.PicName] = source.StartFrame;
                }
            }

            var frameMapping = new List<int>();
            foreach (Match frameMatch in framePattern.Matches(tableContent))
            {
                var picName = frameMatch.Groups[1].Value;
                var frameIndexInPng = int.Parse(frameMatch.Groups[2].Value);

                // Adjust frame index based on which PNG this came from
                var offset = picToOffset.GetValueOrDefault(picName, 0);
                frameMapping.Add(offset + frameIndexInPng);
            }

            // Handle obj_frame_tiles (single frame sprites)
            if (frameMapping.Count == 0 && tableContent.Contains("obj_frame_tiles"))
            {
                frameMapping.Add(0);
            }

            if (frameMapping.Count > 0)
            {
                result[spriteName] = frameMapping;
            }
        }

        return result;
    }
}

/// <summary>
/// Parsed animation data from pokeemerald source.
/// </summary>
public class AnimationData
{
    public Dictionary<string, string> PicToFilePath { get; set; } = new();
    public Dictionary<string, List<string>> MultiFilePics { get; set; } = new();
    public Dictionary<string, List<AnimFrame>> AnimationSequences { get; set; } = new();
    public Dictionary<string, List<SpriteAnimationDefinition>> AnimationTables { get; set; } = new();
    public Dictionary<string, string> SpriteToAnimTable { get; set; } = new();
    public Dictionary<string, List<SpriteSourceInfo>> PicTableSources { get; set; } = new();
    public Dictionary<string, int> FrameCounts { get; set; } = new();
    public Dictionary<string, List<int>> FrameMappings { get; set; } = new();
}

/// <summary>
/// Single frame in an animation sequence.
/// </summary>
public class AnimFrame
{
    public int FrameIndex { get; set; }
    public int Duration { get; set; } // In GBA ticks (~60fps)
    public bool FlipHorizontal { get; set; }
}

/// <summary>
/// Animation definition with name and frames.
/// </summary>
public class SpriteAnimationDefinition
{
    public string Name { get; set; } = "";
    public List<AnimFrame> Frames { get; set; } = new();
}

/// <summary>
/// Information about a sprite source file.
/// </summary>
public class SpriteSourceInfo
{
    public string PicName { get; set; } = "";
    public string FilePath { get; set; } = "";
    public int StartFrame { get; set; }
    public int FrameCount { get; set; }
}
